Teach Azure Filesystem to authenticate using DefaultAzureCredential in the Python SDK

creste opened the pull request:
Fixes #24210.
Partially implements #20511.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[![Build python source distribution and wheels](https://github.com/apache/beam/workflows/Build%20python%20source%20distribution%20and%20wheels/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[![Python tests](https://github.com/apache/beam/workflows/Python%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[![Java tests](https://github.com/apache/beam/workflows/Java%20Tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[![Go tests](https://github.com/apache/beam/workflows/Go%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

creste commented:
R: @Abacn

Abacn commented:
Thanks @creste. Left some comments below the issue page. For the tests, to fix Lint and Formatter error, one could do
```
# Run from root beam repo dir
pip install yapf==0.29.0
git diff master --name-only | grep "\.py$" | xargs yapf --in-place
```
or
```
# Run from sdks/python
tox -e py3-yapf
```
For RAT error, adding apache license at the top of new files (see other source files in the project)

Abacn commented on sdks/python/apache_beam/io/azure/integration_test/create_certificate.sh:
```diff
@@ -0,0 +1,2 @@
+#!/bin/bash
```
e.g. We still need Apache License header here. And better move the raw .pem files here and write them to local at test run time as part of this script.

creste replied:
I deleted `create_certificate.sh` and moved the cert creation code to `azure_integration_test.sh`, then I deleted `cert.pem` and `key.pem` to avoid the license header issues.

Generating the certificate on every test run adds just a second or two to the total test time.

I hope that solution is acceptable, but happy to change it if not.

creste commented:
> ```
> # Run from root beam repo dir
> pip install yapf==0.29.0
> git diff master --name-only | grep "\.py$" | xargs yapf --in-place
> ```

Thank you for the tip! I ran those commands and fixed all linting errors in the files I modified.  One exception is `sdks/python/setup.py`, which appears to have many style issues outside of my changes.  I didn't fix those.

The PythonLint Jenkins job is [failing](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Commit/16635/console) with linting errors outside of the code I modified:
> 18:47:35 ************* Module apache_beam.typehints.batch
> 18:47:35 apache_beam/typehints/batch.py:32:0: W0611: Unused List imported from typing (unused-import)
> 18:47:35 ************* Module apache_beam.dataframe.frames
> 18:47:35 apache_beam/dataframe/frames.py:679:53: I1101: Module 'pandas._libs.lib' has no 'no_default' member, but source is unavailable. Consider adding this module to extension-pkg-allow-list if you want to perform analysis based on run-time introspection of living objects. (c-extension-no-member)

I'm not sure how to proceed.

Abacn commented:
> > ```
> > # Run from root beam repo dir
> > pip install yapf==0.29.0
> > git diff master --name-only | grep "\.py$" | xargs yapf --in-place
> > ```
> 
> Thank you for the tip! I ran those commands and fixed all linting errors in the files I modified. One exception is `sdks/python/setup.py`, which appears to have many style issues outside of my changes. I didn't fix those.
> 
> The PythonLint Jenkins job is [failing](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Commit/16635/console) with linting errors outside of the code I modified:
> 
> > 18:47:35 ************* Module apache_beam.typehints.batch
> > 18:47:35 apache_beam/typehints/batch.py:32:0: W0611: Unused List imported from typing (unused-import)
> > 18:47:35 ************* Module apache_beam.dataframe.frames
> > 18:47:35 apache_beam/dataframe/frames.py:679:53: I1101: Module 'pandas._libs.lib' has no 'no_default' member, but source is unavailable. Consider adding this module to extension-pkg-allow-list if you want to perform analysis based on run-time introspection of living objects. (c-extension-no-member)
> 
> I'm not sure how to proceed.

The remaining linting error is 
```
18:47:35 ************* Module apache_beam.typehints.batch
18:47:35 apache_beam/typehints/batch.py:32:0: W0611: Unused List imported from typing (unused-import)
```
not relevant to the change. It was introduced in #24022 but somehow not detected by linter. Never mind.

Abacn commented:
https://github.com/apache/beam/blob/b952b41788acc20edbe5b75b2196f30dbf8fdeb0/build.gradle.kts#L342

adding `dependsOn(":sdks:python:test-suites:direct:py37:azureIntegrationTest")` below to add it into postcommit test suite. Running with Py37 should suffice.

Abacn commented:
Run Python 3.7 PostCommit

Abacn commented:
integration test passed:
```
13:28:40 test_1     |   azure_integration_test: commands succeeded
13:28:40 test_1     |   congratulations :)
13:28:59 azure_it-python_3_7-jenkins-beam_postcommit_python37_pr-493_test_1 exited with code 0
13:28:59 Stopping azure_it-python_3_7-jenkins-beam_postcommit_python37_pr-493_azurite_1 ... 
13:29:00 Stopping azure_it-python_3_7-jenkins-beam_postcommit_python37_pr-493_azurite_1 ... done
13:29:00 Aborting on container exit...
13:29:00 
13:29:00 real	1m39.973s
13:29:00 user	0m1.198s
13:29:00 sys	0m0.143s
13:29:00 + finally
13:29:00 + docker-compose -p azure_IT-python_3_7-jenkins-beam_PostCommit_Python37_PR-493 --no-ansi down
13:29:01 Removing azure_it-python_3_7-jenkins-beam_postcommit_python37_pr-493_test_1    ... 
13:29:01 Removing azure_it-python_3_7-jenkins-beam_postcommit_python37_pr-493_azurite_1 ... 
13:29:02 Removing azure_it-python_3_7-jenkins-beam_postcommit_python37_pr-493_test_1    ... done
13:29:02 Removing azure_it-python_3_7-jenkins-beam_postcommit_python37_pr-493_azurite_1 ... done
13:29:02 Removing network azure_it-python_3_7-jenkins-beam_postcommit_python37_pr-493_azure_test_net
13:29:02 
13:29:02 real	0m1.785s
13:29:02 user	0m0.900s
13:29:02 sys	0m0.248s
```

Abacn reviewed (approved):
Looks pretty solid to me. The integration test follows the same pattern of hdfs integration test we currently have. CC: @pabloem who reviewed blobstorageio if having any inputs.

Abacn commented:
ah I see, CHANGES will go to upcoming 2.44.0 here: https://github.com/apache/beam/blob/026563483809edb7d7e9873b729830d7d857e41e/CHANGES.md?plain=1#L63. Need a rebase to latest master.

creste commented:
@Abacn - Oh, thanks! I [fixed CHANGES.md](https://github.com/apache/beam/pull/24212/commits/cb171a416005cd90e272bd90e07937c45ba8c066) and rebased onto the latest master.
